### ISSUE TITLE
Memory Leak when I am using tagger.predict() on batch sentences 

### ISSUE THREAD

**Issue by lucky050619** [flairNLP/flair#1639]
I have trained a custom ner model and trying to predict the labels in a loop. 

 trained the model with  stacked embedding model with flair-frwd ,flair-back, byte-pair and glove. 

it is adding memory to RAM for every predict method.

versions:
torch==1.3.1
flair==0.4.3

           2408.3 MiB      0.0 MiB           
           2408.3 MiB      0.0 MiB           
           2449.3 MiB     41.1 MiB           self.model.predict(metadata_sentences, mini_batch_size=64)

**Comment by alanakbik** [comment]
Are you using normal FlairEmbeddings or the pooled variant?

**Comment by lucky050619** [comment]
I am using normal embeddings only. 

embedding_types: List[TokenEmbeddings] = [
        BytePairEmbeddings('en'),
        FlairEmbeddings('news-forward'),
        FlairEmbeddings('news-backward')
    ]
    embeddings: StackedEmbeddings = StackedEmbeddings(embeddings=embedding_types)

![rsz_1screenshot_from_2020-05-27_19-25-56](https://user-images.githubusercontent.com/30066490/83029549-6a54ed80-a050-11ea-857c-430cb55ea3bb.png)

**Comment by alanakbik** [comment]
In this case it should not add memory beyond the memory needed to keep the tags that you predict. Can you share the full script?

**Comment by lucky050619** [comment]
@alanakbik Script looks like this for inference.

class FlairTrainedModels:
    __flair_trained_model = None
    @classmethod
    def get_flair_trained_model(cls):
        if cls.__flair_trained_model is None:
            cls.__flair_trained_model = SequenceTagger.load('best-model.pt')
        return cls.__flair_trained_model



class FlairModelPrediciton:

    def __init__(self):
        self.model = FlairTrainedModels.get_flair_trained_model()
    def flair_inference(self, metadata_sentences):
        flair_ner_results = []
        metadata_sentences = [Sentence(sent) for sent in metadata_sentences]
        self.model.predict(metadata_sentences, mini_batch_size=64)
        for sente in metadata_sentences:
            model_result = sente.to_dict(tag_type='ner')
            flair_ner_results.append(model_result)
        return flair_ner_results

**Comment by alanakbik** [comment]
Could it be that the memory buildup is the results being stored in the array that you append to?

**Comment by lucky050619** [comment]
@alanakbik Memory Profiler looks like this. 

![Screenshot from 2020-05-28 16-44-32](https://user-images.githubusercontent.com/30066490/83135184-19ea9800-a103-11ea-9055-33dfaafad47e.png)

**Comment by vaibhavbhalla07** [comment]
@alanakbik  Even I am facing this issue. 

After analyzing during inference, the exact memory addition is at language_model.py line 46

at line-->    nn.LSTM(embedding_size, hidden_size, layers)

where nn is from import torch.nn as nn

![image](https://user-images.githubusercontent.com/9915983/84761104-92e45d80-afe6-11ea-8a36-aa7c06754897.png)


Now I guess this is the problem with Torch just needed your views on this.

Thanks in advance.

**Comment by blodstone** [comment]
I found out that if I don't manually clear the embedding, the object persists in GC and cause OOM.
`    def forward(self, tokens: List[List]) -> torch.Tensor:
        length = max([len(line.split()) for line in tokens])
        sents_flairs = [Sentence(sentence, use_tokenizer=False) for sentence in tokens]
        FlairTokenEmbedder.bert_embedding.embed(sents_flairs)
        self.hidden_dim = sents_flairs[0].tokens[0].embedding.shape[0]
        embedding = torch.zeros((len(tokens), length, self.hidden_dim), device=self.device)
        for i in range(len(tokens)):
            for j, token in enumerate(sents_flairs[i]):
                embedding[i, j, :] = token.embedding.clone()
        for sent in sents_flairs:
            for token in sent:
                token.clear_embeddings() # Without this, there will be OOM issue
        return embedding`

Is this an expected behaviour or is a bug?
